src/printer.rs

```diff
@@ -0,0 +1,299 @@
+    if value.len() < 5 {
+      return None;
+    }
+    // 5 is length of `jsr:@`/`npm:@`
+    let Some(at_index) = value[5..].find('@').map(|i| i + 5) else {
+      // no version requirement
+      // ex. `npm:jsonc-parser` or `jsr:@pkg/scope`
+      return Some((value, None));
+    };
+    let name = &value[..at_index];
+    let version = &value[at_index + 1..];
+    Some((name, Some(version)))
+  }
+
+  let mut pkg_had_multiple_specifiers: HashMap<&str, bool> =
+    HashMap::with_capacity(specifiers.len());
+  for req in specifiers.keys() {
+    let Some((name, _)) = split_pkg_req(req) else {
+      continue; // corrupt
+    };
+    pkg_had_multiple_specifiers
+      .entry(name)
+      .and_modify(|v| *v = true)
+      .or_default();
+  }
+
+  output.push_str(",\n  \"jsr\": {\n");
+  for (i, (key, value)) in jsr.iter().enumerate() {
+    if i > 0 {
+      output.push_str(",\n");
+    }
+    output.push_str("    \"");
+    output.push_str(key);
+    output.push_str("\": {\n");
+    output.push_str("      \"integrity\": \"");
+    output.push_str(&value.integrity);
+    output.push('"');
+    if !value.dependencies.is_empty() {
+      output.push_str(",\n      \"dependencies\": [\n");
+      for (i, dep) in value.dependencies.iter().enumerate() {
+        if i > 0 {
+          output.push_str(",\n");
+        }
+        output.push_str("        \"");
+        // todo: don't unwrap here
+        let (name, _req) = split_pkg_req(dep).unwrap();
+        let has_single_specifier = pkg_had_multiple_specifiers
+          .get(name)
+          .map(|had_multiple| !had_multiple)
+          .unwrap_or(false);
+        if has_single_specifier {
+          output.push_str(name);
+        } else {
+          output.push_str(dep);
+        }
+
+        output.push('"');
+      }
+      output.push_str("\n      ]");
+    }
+    output.push_str("\n    }");
+  }
+  output.push_str("\n  }");
+}
+
+fn write_npm(output: &mut String, npm: &BTreeMap<String, NpmPackageInfo>) {
+  fn extract_nv_from_id(value: &str) -> Option<(&str, &str)> {
+    if value.is_empty() {
+      return None;
+    }
+    let at_index = value[1..].find('@').map(|i| i + 1)?;
+    let name = &value[..at_index];
+    let version = &value[at_index + 1..];
+    Some((name, version))
+  }
+
+  let mut pkg_had_multiple_versions: HashMap<&str, bool> =
+    HashMap::with_capacity(npm.len());
+  for id in npm.keys() {
+    let Some((name, _)) = extract_nv_from_id(id) else {
+      continue; // corrupt
+    };
+    pkg_had_multiple_versions
+      .entry(name)
+      .and_modify(|v| *v = true)
+      .or_default();
+  }
+
+  output.push_str(",\n  \"npm\": {\n");
+  for (i, (key, value)) in npm.iter().enumerate() {
+    if i > 0 {
+      output.push_str(",\n");
+    }
+    output.push_str("    \"");
+    output.push_str(key);
+    output.push_str("\": {\n");
+    output.push_str("      \"integrity\": \"");
+    output.push_str(&value.integrity);
+    output.push('"');
```

Comment:
Is it really better to hand write JSON in this file?

Reply:
Not sure. I wanted to avoid clones and haven't meadured the performance of this yet. I'm going to rewrite this once dropping v3 I think.